### PR TITLE
Disable autocompaction in tests

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**: Our ETE tests occasionally fail because a compaction is kicked off and
everything else grinds to a halt. Disabling compaction in this way
should resolve the issue.

Example failing build:
https://circleci.com/gh/palantir/atlasdb/12040#tests/containers/1

**Implementation Description (bullets)**: Use our nodetool hook to disable compaction via JMX

**Priority (whenever / two weeks / yesterday)**: medium priority, not release blocking but would be nice to get this in and see if it finally squashes the heartbeat flake.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2566)

### CODE REVIEW

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java [hunk 1]
```diff
@@ -87,11 +93,29 @@ private String runCql(String cql) throws IOException, InterruptedException {
...
     private String runNodetoolCommand(String nodetoolCommand, int timeoutSeconds) throws IOException,
             InterruptedException {
+        return runNodetoolCommandOnContainer(nodetoolCommand, timeoutSeconds,
+                ThreeNodeCassandraCluster.FIRST_CASSANDRA_CONTAINER_NAME);
+    }
+
+    private List<String> runNodetoolCommandOnAllContainers(String nodetoolCommand, int timeoutSeconds)
+            throws IOException, InterruptedException {
+        List<String> ret = Lists.newArrayList();
+        ret.add(runNodetoolCommandOnContainer(nodetoolCommand, timeoutSeconds,
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: Use a stream/loop here after adding a constant for `List<CassandraContainers>`.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java [hunk 2]
```diff
@@ -58,6 +60,10 @@ public void replicateSystemAuthenticationDataOnAllNodes()
         }
     }
 
+    public void disableAutoCompaction() throws IOException, InterruptedException {
+        runNodetoolCommandOnAllContainers("disableautocompaction", NODETOOL_REPAIR_TIMEOUT_SECONDS);
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: Interesting, is this a problem only when multinode tests are running? If not should we run this command on the docker image https://hub.docker.com/r/palantirtechnologies/docker-cassandra-atlasdb/~/dockerfile/ add a line `CMD ["nodetool disableautocompaction"]`.?

Comment (tpetracca) [inline atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: would suggest a much much smaller timeout. I'd expect this to be faster than `nodetool status`.

Comment (tpetracca) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: +1 to Himangi's question

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java (hunk 1 above)
Comment (tpetracca) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: again +1; an ImmutableList would also work.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java (hunk 2 above)
Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: a) the suggested command failed :(
b) This PR disproved that it's only a multinode problem - a build failed because of the heartflake on a single-node container!

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java (hunk 1 above)
Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/ThreeNodeCassandraClusterOperations.java]: Done. Stream didn't work, as `runNodetoolCommandOnContainer` throws checked exceptions.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraClusterOperations.java [hunk 3]
```diff
@@ -81,25 +97,39 @@ private String runCql(String cql) throws IOException, InterruptedException {
...
+                containerNames.get(0),
                 "--execute", cql);
     }
 
     private String runNodetoolCommand(String nodetoolCommand, int timeoutSeconds) throws IOException,
             InterruptedException {
+        return runNodetoolCommandOnContainer(nodetoolCommand, timeoutSeconds, containerNames.get(0));
+    }
+
+    private List<String> runNodetoolCommandOnAllContainers(String nodetoolCommand, int timeoutSeconds)
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraClusterOperations.java]: Shouldnt this return void as the return value is not used anywhere. I think the intent was to follow `runNodetoolCommandOnContainer`, but this doesnt look necessary here.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java [hunk 4]
```diff
@@ -75,9 +79,26 @@ public String getDockerComposeFile() {
...
     @Override
     public SuccessOrFailure isReady(DockerComposeRule rule) {
-        return SuccessOrFailure.onResultOf(() -> CassandraKeyValueServiceImpl.create(
+        return SuccessOrFailure.onResultOf(() -> {
+            CassandraClusterOperations cassandraOperations = new CassandraClusterOperations(rule,
+                    CASSANDRA_VERSION,
+                    ImmutableList.of(CONTAINER_NAME),
+                    CLI_CONTAINER_NAME);
+
+            if (!cassandraOperations.nodetoolShowsAllCassandraNodesUp()) {
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java]: This is a change of behavior. Some tests might not require all nodes to be up and trying to enforce that here seems like a stricter check. The earlier check `createKVS` can succeed if we have quorum too. We can perhaps have a classRule in the test classes to disableAutoCompaction if thats the first thing to do? Does this nodetool command require the node to be up first?

File: atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra.yml [hunk 5]
```diff
@@ -7,3 +7,10 @@ services:
       - MAX_HEAP_SIZE=$CASSANDRA_MAX_HEAP_SIZE
       - HEAP_NEWSIZE=$CASSANDRA_HEAP_NEWSIZE
       - LOCAL_JMX=no
+
+  cli:
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra.yml]: Why was this necessary?

Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra.yml]: The nodetool command is run on a separate container by default - I maintained that convention, but we hadn't needed to do so with the single-node setup, so that meant adding this container. Would it be safe to run the commands in the `cassandra` container instead?

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java (hunk 4 above)
Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java]: Note that there's only one node in the cluster, so quorum is the same as 'all' here - so in which case would `createKVS` succeed where getting nodetool status fails?

It seems to me that there would just be a race condition where nodetool status would not show all nodes as up, but in the same iteration, `createKVS` succeeds - and in this case, we'd succeed a second later anyway. Attempting to divine which classes need to have `disableAutoCompaction` set would lead to more complex code and we might not catch all cases where the tests would lead to compaction.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraClusterOperations.java (hunk 3 above)
Comment (gsheasby) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraClusterOperations.java]: We can return void here.

File: atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra.yml (hunk 5 above)
Comment (hsaraogi) [reply atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra.yml]: I think we can run it on the cassandra container but wont push on it if its significant lift.

File: atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java (hunk 4 above)
Comment (hsaraogi) [reply atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/CassandraContainer.java]: Cool, lets leave it as is.

**Comment by gsheasby** [comment]
It doesn't :( the latest failure on internal-circle (build 7228) is due to the heartflake.

Some data points about this (taking the safe-sounding assumption that when Cassandra writes to `Memtable-compactions_in_progress`, it's doing a compaction):
- compactions did occur on build 7228
- compactions _also_ occurred on at least one recent passing build (7225)
- it still may be that compactions make the failure scenario (the lock getting stuck) more likely - I have yet to prove or disprove a link between the 7228 failure and compaction slowing things down.

This comment on [CASSANDRA-5074](https://issues.apache.org/jira/browse/CASSANDRA-5074?focusedCommentId=13836822&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13836822) could be related after all - we disable autocompaction before creating the Atlas tables, so perhaps we want to make the call later.

I may not have bandwidth to investigate further in the near future, unless the heartflake becomes more painful for people. Leaving the PR open for now.

**Comment by hsaraogi** [comment]
@gsheasby should we keep this open?

**Comment by gsheasby** [comment]
No, I'll file an issue for the flaky test + investigation instead.

**Comment by gsheasby** [comment]
Closing - filed #2666.
